File: src/main/java/pl/training/module2/SeamedReminderService.java
package pl.training.module2;

import java.time.LocalDate;
import java.time.ZoneOffset;

public class SeamedReminderService {
    public boolean sendRenewalReminder(Subscription subscription) {
        LocalDate today = currentDate();

        if (subscription.renewalDate().isAfter(today.plusDays(7))) {
            return false;
        }

        sendMessage(subscription.email(), subscription.renewalDate());
        return true;
    }

    protected LocalDate currentDate() {
        return LocalDate.now(ZoneOffset.UTC);
    }

    protected void sendMessage(String email, LocalDate renewalDate) {
        System.out.printf(
                "Sent renewal reminder to %s for %s%n",
                email,
                renewalDate);
    }
}

File: typescript/src/module2/SeamedReminderService.ts
import { LocalDate } from './LocalDate.js';
import type { Subscription } from './Subscription.js';

// Klasa celowo otwarta na dziedziczenie: chronione metody są szwami (seams).
export class SeamedReminderService {
  sendRenewalReminder(subscription: Subscription): boolean {
    const today = this.currentDate();

    if (subscription.renewalDate.isAfter(today.plusDays(7))) {
      return false;
    }

    this.sendMessage(subscription.email, subscription.renewalDate);
    return true;
  }

  protected currentDate(): LocalDate {
    return LocalDate.now();
  }

  protected sendMessage(email: string, renewalDate: LocalDate): void {
    console.log(`Sent renewal reminder to ${email} for ${renewalDate.toString()}`);
  }
}
